VersionControlUI/src/DiffManager.cpp

```diff
@@ -0,0 +1,288 @@
+
+		Model::NodeIdType nodeId;
+
+		// TODO maybe use node instead of id
+		if (change->type() != FilePersistence::ChangeType::Deletion)
+			if (findChangedNode(newVersionManager, id, nodeId))
+				changedNodesToVisualize.insert(nodeId);
+		if (change->type() != FilePersistence::ChangeType::Insertion)
+			if (findChangedNode(oldVersionManager, id, nodeId))
+				changedNodesToVisualize.insert(nodeId);
+	}
+
+	removeDirectChildrenOfNodesInContainer(&changesWithNodes);
+
+	Visualization::VisualizationManager::instance().mainScene()->listenToTreeManager(newVersionManager);
+	Visualization::VisualizationManager::instance().mainScene()->listenToTreeManager(oldVersionManager);
+
+	auto diffViewItem = Visualization::VisualizationManager::instance().mainScene()->
+			viewItems()->newViewItem("DiffView");
+
+	diffViewItem->setMajorAxis(Visualization::GridLayouter::NoMajor);
+
+	// add nodes to visualization, if not avaible show text
+	visualizeChangedNodes(oldVersionManager, changedNodesToVisualize, newVersionManager, diffViewItem);
+
+	// create visualization for changes
+	createOverlaysForChanges(diffViewItem, changesWithNodes);
+
+	// switch to the newly created view
+	Visualization::VisualizationManager::instance().mainScene()->viewItems()->switchToView(diffViewItem);
+}
+
+// TODO maybe better implementation for this
+void DiffManager::removeDirectChildrenOfNodesInContainer(QList<ChangeWithNodes>* container)
+{
+	QHash<Model::Node*, FilePersistence::ChangeType> oldNodes;
+	QHash<Model::Node*, FilePersistence::ChangeType> newNodes;
+
+	for (auto change : *container)
+	{
+		if (change.oldNode_)
+			oldNodes.insert(change.oldNode_, change.changeType_);
+		if (change.newNode_)
+			newNodes.insert(change.newNode_, change.changeType_);
+	}
+
+	QSet<Model::Node*> oldNodesMarkedForRemoval = findAllNodesWithDirectParentPresent(oldNodes);
+	QSet<Model::Node*> newNodesMarkedForRemoval = findAllNodesWithDirectParentPresent(newNodes);
+
+	auto it = container->begin();
+	while (it != container->end())
+	{
+		if (oldNodesMarkedForRemoval.contains(it->oldNode_) || newNodesMarkedForRemoval.contains(it->newNode_))
+			it = container->erase(it);
+		else
+			it++;
+	}
+}
+
+QSet<Model::Node*> DiffManager::findAllNodesWithDirectParentPresent(QHash<Model::Node*,
+																						  FilePersistence::ChangeType>& nodes)
+{
+	QSet<Model::Node*> nodesMarkedForRemoval;
+	for (auto node : nodes.keys())
+		// check that changeType also matches
+		if (nodes.contains(node->parent()) && nodes[node] == nodes[node->parent()])
+				nodesMarkedForRemoval.insert(node);
+
+	return nodesMarkedForRemoval;
+}
+
+// TODO find good way to return and use Node instead of Id
+bool DiffManager::findChangedNode(Model::TreeManager* treeManager, Model::NodeIdType id, Model::NodeIdType& resultId)
+{
+	if (auto node = const_cast<Model::Node*>(treeManager->nodeIdMap().node(id)))
+	{
+
+		Model::Node* changedNode = nullptr;
+		Model::SymbolMatcher symbolMatcher{symbol_};
```

Comment:
Hmm, instead of saving `QString symbol_` field, you should directly take a `SymbolMatcher` in the constructor. That way this code here is more efficient, and the client has more options to choose what is matched.